extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,580 @@
+   * @returns {!Object<string, number>}
+   */
+  getExpandBounds(changes) {
+    const frameHeight = changes.boundingClientRect.bottom -
+          changes.boundingClientRect.top;
+    const frameWidth = changes.boundingClientRect.right -
+          changes.boundingClientRect.left;
+    const viewportHeight = changes.rootBounds.bottom -
+          changes.rootBounds.top;
+    const viewportWidth = changes.rootBounds.right -
+          changes.rootBounds.left;
+    return {
+      width: (viewportWidth - frameWidth) / 2,
+      height: (viewportHeight - frameHeight) / 2,
+    };
+  }
+
+  /**
+   * Handles serializing and sending messages to the safeframe.
+   * @param {!Object} payload
+   * @param {string} serviceName
+   * @private
+   */
+  sendMessage_(payload, serviceName) {
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    const message = {};
+    message[MESSAGE_FIELDS.CHANNEL] = this.channel;
+    message[MESSAGE_FIELDS.PAYLOAD] = JSON.stringify(payload);
+    message[MESSAGE_FIELDS.SERVICE] = serviceName;
+    message[MESSAGE_FIELDS.SENTINEL] = this.sentinel_;
+    message[MESSAGE_FIELDS.ENDPOINT_IDENTITY] = this.endpointIdentity_;
+    this.iframe_.contentWindow./*OK*/postMessage(
+        JSON.stringify(message), SAFEFRAME_ORIGIN);
+  }
+
+  /**
+   * Routes messages to their appropriate handler.
+   * @param {!Object} payload
+   * @param {string} service
+   */
+  processMessage(payload, service) {
+    switch (service) {
+      case SERVICE.CREATIVE_GEOMETRY_UPDATE:
+        this.handleFluidMessage_(payload);
+        break;
+      case SERVICE.EXPAND_REQUEST:
+        this.handleExpandRequest_(payload);
+        break;
+      case SERVICE.REGISTER_DONE:
+        this.handleRegisterDone_(payload);
+        break;
+      case SERVICE.COLLAPSE_REQUEST:
+        this.handleCollapseRequest_();
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * @param {!Object} payload
+   * @private
+   */
+  handleRegisterDone_(payload) {
+    this.initialHeight_ = payload['initialHeight'];
+    this.initialWidth_ = payload['initialWidth'];
+  }
+
+  /**
+   * @param {!Object} payload
+   * @private
+   */
+  handleExpandRequest_(payload) {
+    let expandHeight = Math.floor(payload.expand_b +
+                                 payload.expand_t);
+    let expandWidth = Math.floor(payload.expand_r +
+                                 payload.expand_l);
+    // We assume that a fair amount of usage of this API will try to pass the final
+    // size of the iframe that is desired, instead of the correct usage which is
+    // specify how much to expand by. We try to protect against this, by detecting
+    // if the requested expansion size matches the size of the amp-ad element, and
+    // if so we assume that the desired effect was to resize to fill the element.
+    if (expandWidth != this.baseInstance_.element.style.width.split('px')[0] ||
```

Comment:
Can this be based on getLayoutRect which returns numeric values?  Although that would be based on getBoundingClientRect info so not sure what you want here.